clang/lib/Driver/ToolChain.cpp

```diff
@@ -1370,19 +1370,40 @@ ToolChain::CXXStdlibType ToolChain::GetCXXStdlibType(const ArgList &Args) const{
   return *cxxStdlibType;
 }
 
+static void ResolveAndAddSystemIncludePath(const ArgList &DriverArgs,
+                                           ArgStringList &CC1Args,
+                                           const Twine &Path) {
+  bool Canonicalize =
+      DriverArgs.hasFlag(options::OPT_canonical_prefixes,
+                         options::OPT_no_canonical_prefixes, true);
+
+  if (!Canonicalize) {
+    CC1Args.push_back(DriverArgs.MakeArgString(Path));
+    return;
+  }
+
+  // We canonicalise system include paths that were added automatically
+  // since those can end up quite long otherwise.
+  SmallString<256> Canonical, PathStorage;
+  StringRef SimplifiedPath = Path.toStringRef(PathStorage);
+  if (!sys::fs::real_path(SimplifiedPath, Canonical))
+    SimplifiedPath = Canonical;
+  CC1Args.push_back(DriverArgs.MakeArgString(SimplifiedPath));
+}
+
 /// Utility function to add a system framework directory to CC1 arguments.
 void ToolChain::addSystemFrameworkInclude(const llvm::opt::ArgList &DriverArgs,
                                           llvm::opt::ArgStringList &CC1Args,
                                           const Twine &Path) {
   CC1Args.push_back("-internal-iframework");
-  CC1Args.push_back(DriverArgs.MakeArgString(Path));
+  ResolveAndAddSystemIncludePath(DriverArgs, CC1Args, Path);
 }
 
 /// Utility function to add a system include directory to CC1 arguments.
 void ToolChain::addSystemInclude(const ArgList &DriverArgs,
                                  ArgStringList &CC1Args, const Twine &Path) {
   CC1Args.push_back("-internal-isystem");
```

Comment:
Perhaps "-internal-isystem" should be an argument to the `addSystemIncludePath` function